### PR TITLE
change(rpc): generate coinbase transactions in the getblocktemplate RPC

### LINKED ISSUE #5453: Populate transparent coinbase transaction data for `BlockTemplate`
## Motivation

The `coinbasetxn`  field of the `BlockTemplate` is a struct:(https://github.com/ZcashFoundation/zebra/issues/5235#issuecomment-1285174731) and can be complex enough to implement it in a separated ticket.

## Specifications

- [x] [NU5 onward] The transaction version number MUST be 4 or 5. If the transaction version number is 4 then the version group ID MUST be 0x892F2085. If the transaction version number is 5 then the version group ID MUST be 0x26A7270A.
- [x] [NU5 onward] If effectiveVersion ≥ 5, the nConsensusBranchId field MUST match the consensus branch ID used for SIGHASH transaction hashes, as specified in [ZIP-244].
- [ ] The total value in zatoshi of transparent outputs from a coinbase transaction, minus vbalanceSapling, minus vbalanceOrchard, MUST NOT be greater than the value in zatoshi of block subsidy plus the transaction fees paid by transactions in this block.
- [x] [NU5 onward] In a version 5 coinbase transaction, the enableSpendsOrchard flag MUST be 0.
- [x] [NU5 onward] In a version 5 transaction, the reserved bits 2 .. 7 of the flagsOrchard field MUST be zero.
- [x] A coinbase transaction for a block at block height greater than 0 MUST have a script that, as its first item, encodes the block height height as follows. For height in the range {1 .. 16}, the encoding is a single byte of value 0x50 + height. Otherwise, let heightBytes be the signed little-endian representation of height, using the minimum nonzero number of bytes such that the most significant byte is < 0x80. The length of heightBytes MUST be in the range {1 .. 5}. Then the encoding is the length of heightBytes encoded as one byte, followed by heightBytes itself. This matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the description here is to be considered normative).
- [x] A coinbase transaction script MUST have length in {2 .. 100} bytes.
- [x] [NU5 onward] The nExpiryHeight field of a coinbase transaction MUST be equal to its block height .

https://zips.z.cash/protocol/protocol.pdf#txnconsensus

## Tasks

Shielded coinbase is out of scope, most miners don't use it.

- Create `CoinBaseTxn` struct.
- Populate on `getblocktemplate` response
  - The coinbase transaction has the following fixed values:
    - `required`: `true`
  - Here are the fields that are calculated from the transactions, including the coinbase transaction:
    * `defaultroots`
      * `authdataroot`
      * `merkleroot`

- Test: TODO

Depends on https://github.com/ZcashFoundation/zebra/issues/5451

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

This PR generates coinbase transactions when the `getblocktemplate` RPC is called.

## Solution

Generate the coinbase transaction in the getblocktemplate RPC:
- Add coinbase creation to zebra-chain
- Add coinbase creation and miner subsidy to zebra-consensus

Related changes:
- Add the miner config to the GetBlockTemplateRpcImpl
- Add a getblocktemplate-rpcs feature to zebra-chain

Related fixes:
- Fix missing feature dependencies in zebra-consensus
- Provide fake valid block heights to getblocktemplate RPC tests
- Rewrite `check_script_form()` so it is easier to read

Testing:
- Add a getblocktemplate.coinbase_tx deserialized transaction snapshot
- Update the getblocktemplate RPC snapshots

Closes #5453.

## Review

This is a routine change, it might help getting two people to review it.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

I split #5579 out of #5453, because the state request wasn't ready yet.

### CODE REVIEW

**Review by oxarbitrage — APPROVED** [review]
Looks really good.

File: zebra-consensus/src/block/subsidy/funding_streams.rs [hunk 1]
```diff
@@ -124,20 +124,43 @@ pub fn funding_stream_address(
 ///
 /// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
 pub fn check_script_form(lock_script: &Script, address: Address) -> bool {
-    let mut address_hash = address
+    // TODO: Verify P2SH multisig funding stream addresses (#5577).
+    //       As of NU5, the funding streams do not use multisig addresses,
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/funding_streams.rs:128]: This doesn't need to be addressed in this PR, but the funding streams might use multisig addresses: https://github.com/ZcashFoundation/zebra/issues/5577#issuecomment-1309292394

File: zebra-consensus/src/block/subsidy/funding_streams.rs [hunk 2]
```diff
@@ -124,20 +124,43 @@ pub fn funding_stream_address(
...
+    //       so this is optional.
```
Comment (str4d) [inline zebra-consensus/src/block/subsidy/funding_streams.rs:129]: This is not the case. All of the (currently-revealed) funding stream addresses are P2SH ([as indicated by their `t3` prefixes](https://github.com/zcash/zcash/blob/fadbb794f16bee26513eb9a1fda51ac0cfa096e1/src/chainparams.cpp#L171-L225)), and of the addresses with publicly-revealed `scriptSig` values, all of the ones I spot-checked are multisig addresses. In particular, the ZF's funding stream address (the `scriptSig` for which is visible in [this transaction](https://blockchair.com/zcash/transaction/866de6abfee9ad02c267dccfc45710d8847b66a72e3b458f120c94ca0b4bf83c?i=999)) is a 2-of-3 multisig address (with an additional storage of 4 bytes of data that I can't identify):

```
// OP_0
00
// PUSHDATA(<SIGNATURE> || SIGHASH_ALL)
48
  3045022100a16d32df92e28624ae194fb4de9e8610aa1932e4e2b370fa7c0ca7dab62d34d102205cc4f9dac56d43e7fd9e3d70816e3aa7f72832774712193bfbc227fb5ffaa0bf
  01
// PUSHDATA(<SIGNATURE> || SIGHASH_ALL)
47
  304402202341788cde33e568cfc388acf36ae9b24c22883c7ad88822e54c8bbe07a4c93202207654bcaa7c8f14403c424c7d681b4c525f779aa314795066a825e921fd4525d4
  01
// PUSHDATA(multisig script)
4c 72
  52 // OP_2
  21 0352d50656796dd79e3e3385c29636d849b8705c9c6d9b86a1717adc38e4a567c7 // PUSHDATA(pubkey)
  21 02f39d0b69c3b8e06030ef9dc295cf8c848f51e398fe85414e33509afdc7f01fb3 // PUSHDATA(pubkey)
  21 027072ba2319ead2b0b387d69d36508f57b397da8d4ac76ea64676ee8a1356dae0 // PUSHDATA(pubkey)
  53 // OP_3
  af // OP_CHECKMULTISIGVERIFY
  04 d5f2a8b6 // PUSHDATA(0xb6a8f2d5)
  75 // OP_DROP
  74 // OP_DEPTH
  00 // OP_0
  87 // OP_EQUAL
```

File: zebra-consensus/src/block/subsidy/funding_streams.rs (hunk 1 above)
Comment (teor2345) [reply zebra-consensus/src/block/subsidy/funding_streams.rs:128]: Thanks, I'll deal with this in ticket #5577.

File: zebra-consensus/src/block/subsidy/funding_streams.rs (hunk 2 above)
Comment (teor2345) [reply zebra-consensus/src/block/subsidy/funding_streams.rs:129]: Thanks, I'll deal with this in ticket #5577.

**Pull request merged**